### ISSUE TITLE
Transport security redirecting proxied https to http

### ISSUE THREAD

**Issue by adamkdean** [dadi/web#116]
Awaiting logs from @feroviktor on this, but it looks like for some reason, the transport security middleware is pushing a 301 even when it shouldn't.

I'm going to do some testing with a few scenarios.

**Comment by adamkdean** [comment]
```
{ host: '*snip*',
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
  'accept-encoding': 'gzip, deflate, sdch, br',
  'accept-language': 'en,en-US;q=0.8,hu;q=0.6',
  'upgrade-insecure-requests': '1',
  'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36',
  'x-forwarded-for': '*snip*',
  'x-forwarded-port': '443',
  'x-forwarded-proto': 'https',
  connection: 'keep-alive' }
```

**Comment by adamkdean** [comment]
Looking at this and running tests today, I think a few changes need to be made to the transport security middleware. I think we should introduce a `transportSecurity` configuration option to the `security` block so that we can define what action we want to take place.

In the original implementation of SSL support, `server.protocol` would be set to `https` and `security.trustProxy` was required in order to support a load balancer ferrying HTTPS to web listening on HTTP.

Now that we support SSL natively and the nature of `server.protocol` has changed, we need to modify the transportSecurity middleware. I suggest like so:

```js
function securityEnabled() {
  var transportSecurity = config.get('security.transportSecurity')
  var protocol = config.get('server.protocol')
  return protocol === 'https' || transportSecurity
}
```

Using this approach, if we're listening for HTTPS requests, then transport security will be enabled by default, otherwise, we'll have to enable it if we want it. We will still need to enable `security.trustProxy` if we want to trust the values specified in X-Forwarded-* headers.

I've ran these changes through some use cases.

> Without proxy

| Web   | TS | Expected  | Result    | Notes |
|-------|----|-----------|-----------|-------|
| HTTP  | ❌ | 200 OK    | 200 OK    | -     |
| HTTP  | ✅ | 301 HTTPS | 301 HTTPS | -     |
| HTTPS | ❌ | 200 OK    | 200 OK    | 1     |
| HTTPS | ✅ | 200 OK    | 200 OK    | -     |

> With proxy

| Web   | Proxy | TS | TP | Expected  | Result    | Notes |
|-------|-------|----|----|-----------|-----------|-------|
| HTTP  | HTTP  | ❌ | ❌  | 200 OK    | 200 OK    | -     |
| HTTP  | HTTP  | ❌ | ✅  | 200 OK    | 200 OK    | -     |
| HTTP  | HTTP  | ✅ | ❌  | 301 HTTPS | 301 HTTPS | -     |
| HTTP  | HTTP  | ✅ | ✅  | 301 HTTPS | 301 HTTPS | -     |
| HTTP  | HTTPS | ❌ | ❌  | 200 OK    | 200 OK    | -     |
| HTTP  | HTTPS | ❌ | ✅  | 301 HTTP  | 301 HTTP  | -     |
| HTTP  | HTTPS | ✅ | ❌  | 301 HTTPS | 301 HTTPS | 2     |
| HTTP  | HTTPS | ✅ | ✅  | 200 OK    | 200 OK    | -     |

1 – Transport security set to true by server.protocol being set to HTTPS  
2 – Will result in a redirect loop, but this is by design. Without trust proxy enabled, request appears as HTTP but requires HTTPS, redirects to HTTPS, etc.

**Comment by adamkdean** [comment]
So, in the case of @feroviktor's issue, the required configuration would be something along the lines of:

```json
{
  "server": {
    "host": "127.0.0.1",
    "port": 80
  },
  "security": {
    "transportSecurity": true,
    "trustProxy": true
  },
  ...etc
}
```

The server section shows that we'll listen on port 80, but the security section shows that we'll trust the X-Forwarded-* headers and that we'll require the proxy is secure.

**Comment by adamkdean** [comment]
@jimlambie what are your thoughts on this?

**Comment by adamkdean** [comment]
I've got a workspace @ https://github.com/adamkdean/dadi-web-ssl-test if interested.

**Comment by adamkdean** [comment]
This is fixed in 2.0

**Comment by adamkdean** [comment]
This ticket is the best.
